### DISCUSSION TITLE
Витрина магазина

### DISCUSSION THREAD

**Discussion by pingvin1308** [cleannetcode/CsharpForBeginers#15] [Новички]
Представьте себе, что будущее наступило и некий автоматизированный механизм умеет наполнять витрину (и лавки) товарами, а также размещать новые лавки.

Необходимо реализовать консольное приложение, которое умеет работать с лавками и товарами.

Возможности программы:
#Работа с витринами
1) Создать новую витрину с названием и объемом. Объем - сколько товаров может поместиться, у каждого товара свой объем.
2) Отредактировать витрину. При смене объема, если товары уже расположены на витрине и их суммарный объем больше нового значения объема витрины, программа не должна позволять это сделать и сообщать ошибку.
3) Удалить витрину. Если товары расположены на витрине, то программа не должна позволять это сделать и сообщать ошибку.
4) В каждую созданную витрину можно добавить товар. При этом если товар не помещается, программа не должна позволять это сделать и сообщать ошибку

#Работа с товарами
1) Создать товар, который можно расположить на витрине.
2) Отредактировать товар
3) Удалить товар

#Модели
Витрина:
+ Идентификатор (порядковый номер, чтобы однозначно идентифицировать витрину)
+ Название
+ Объем
+ Дата создания
+ Дата удаления

Товар:
+ Идентификатор (порядковый номер, чтобы однозначно идентифицировать товар)
+ Название
+ Занимаемый объем

Товары на витрине:
+ Идентификатор (порядковый номер, чтобы однозначно идентифицировать товар на витрине)
+ Идентификатор витрины (порядковый номер, чтобы однозначно идентифицировать витрину)
+ Идентификатор товара (порядковый номер, чтобы однозначно идентифицировать товар)
+ Кол-во товара - в целом можно и сразу считать объем, если посчитаешь это удобным
+ Стоимость товара

В качестве товаров может быть все что угодно (можно пофантазировать)

В программе должен быть интерфейс взаимодействия с пользователем в виде простого консольного меню. При выполнении операций программа не должна закрываться, для выхода из программы следует сделать отдельную операцию. При выполнении задания нужно стараться использовать интерфейсы для каждой операции, чтобы в последствии реализацию операции можно было легко подменить.

Пример: IPlaceProduct - с методом Place(Product product);
реализация может быть PlaceProduct - который будет размещать товар в коллекции в памяти (в лист)
реализация может быть PlaceProductToFile - который будет размещать товар в памяти и сохранять его в файлике. 
Работу с файликом делать не нужно, про это еще будет занятие. Для начала поработайте с данными в памяти. Нужно что у вас все прочно осело в голове.

**Comment by emedvedeva** [comment]
> Представьте себе, что будущее наступило и некий автоматизированный механизм умеет наполнять витрину (и лавки) товарами, а также размещать новые лавки.

А что такое лавка? В дальнейшем в ТЗ про нее ни слова

**Reply by Z1pP** [reply]
прилавок

**Reply by pingvin1308** [reply]
наверное плохо вводить множественный термин, но я тут это сделал намеренно, чтобы вас запутать. Такое вы увидите в реальной жизни к сожалению. Лавка, прилавок, витрина считайте это одной и той же сущностью))

**Comment by Z1pP** [comment]
https://github.com/Z1pP/HuntTheWumpus
Привет все. Роман, если у вас будет свободная минута, посмотрите пожалуйста дз. Я не могу понять, куда впихнуть проверку на пересечение массива

**Reply by pingvin1308** [reply]
Можно попробовать сделать проверки тут. Перед тем как двигать игрока по полю.

switch (keyInfo.Key)
                {
                    case ConsoleKey.UpArrow:
                        {
                            user.Move(Direction.Up);

                            break;
                        }
                    case ConsoleKey.DownArrow:
                        {
                            user.Move(Direction.Down);
                            break;
                        }
                    case ConsoleKey.RightArrow:
                        {
                            user.Move(Direction.Right);
                            break;
                        }
                    case ConsoleKey.LeftArrow:
                        {
                            user.Move(Direction.Left);
                            break;
                        }
                }

**Reply by Z1pP** [reply]
Сделал приватный метод в классе PlayMap, который проверяет и фиксит перемещения.Это делается уже после рендеринга карты, но еще до добавления объекта на карту. Как мне кажется - очень костыльно сделал.
Вчера думал, как реализовать проверку координат объектов на карте. Что если сделать метод, в который передавать объекты классов (Игрок, вампус, мышь и яма), а внутри этого метода уже организовывать логику...

**Reply by pingvin1308** [reply]
Тут бесчисленное кол-во возможных решений)
К примеру если координаты хранятся у объектов, можно передавать все объекты, которые находятся на расстоянии одной клетки от игрока в метод, который уже проверит какие это объекты и выведет сообщения.
Можно конечно просто передать все объекты карты в этот метод, а там уже все обработать и вывести сообщения.

Если координаты хранятся не в игроке, а в комнате, то можно попробовать играться комнатами. Я последнее время думаю об этой реализации

**Comment by Z1pP** [comment]
https://github.com/Z1pP/CsharpForBeginners/tree/master/Shop
Посидел вечерок за заданием. Прошу не жалеть, критиковать и ругать как только угодно, но по факту
Прочитаю все и приму к сведению. Заранее спасибо!

**Reply by pingvin1308** [reply]
Не жалел :)

**Reply by Z1pP** [reply]
Спасибо, буду исправлять

**Reply by trulander** [reply]
Привет, я поиграл с твоей программой немного и заметил что функция вывода списка товаров у тебя не работает, она как будтое ничего не делает. 
Еще такой момент юзабилити в программе не хватает первичного вывода всех имеющихся товаров по прилавкам, их как я понимаю можно вывести только открывая каждый прилавок отдельно и вызывая метод показать товары который не работает:-D.

**Comment by trulander** [comment]
https://github.com/trulander/CsharpEducation/tree/main/ShowCase
Привет всем. 
Что-то тяжко далась эта домашка, замучался с этими интерфейсами и дженериками)), я по-моему только все запутал из за них.
Интерфейсы постоянно мешали, дженерики не давали мне использовать нормально методы:-D
Про интерфейсы я что понял, так это то что в них не нужно пихать все методы подряд, так как в них можно обьявить только публичные методы, а приватные нужно делать с телом метода, я с этим моментом не разобрался еще как с приватными методами работать.

А дженерики у меня на них были большие планы, и они обрушились на не возможности их воплотить, они оказались не такими гибкими как показалось на первый взгляд, а обьявление дженерик метода в интерфейсе который подразумевается использоваться в разных классах, делает обьявление такого класса кошмаром)), по крайне мере у меня получилось именно так. Я мог исправить ересь с дженериком и просто от него отказаться, но тогда в домашке я бы нигде в итоге его не смог применить)).

Сделал программу почти полностью в графическом исполнении, пока делал столкнулся еще с забавным багом работы с консолью, да он меня заколебал)). Если использовать Console.TreatControlCAsInput = true; то последующий вызов Console.ReadLine() пишет данные как будтое в двойной буфер и на каждый вызов чтения из консоли в код отправляет предыдущие введенные данные.

**Reply by pingvin1308** [reply]
Любопытное использование дженериков. Тяжко было разбираться как все устроено и как все работает.
Зачем я такие сложные домашки стал давать xD

**Comment by coder1coder** [comment]
Домашка "Управление магазином"
[https://github.com/coder1coder/Shop](https://github.com/coder1coder/Shop)

_upd: добавил реализацию репозитория товаров_
_upd: добавил реализацию репозитория витрин_

**Reply by trulander** [reply]
Классно меню реализовал, только вот для того чтобы выполнить команду ты парсишь строку в RoutTo,  по мне лучше использовать enum структуру)

**Reply by coder1coder** [reply]
> Классно меню реализовал, только вот для того чтобы выполнить команду ты парсишь строку в RoutTo,  по мне лучше использовать enum структуру)

Времени совсем не хватило на этой неделе. Основной упор делал все таки на более менее рабочий вариант с дружелюбным интерфейсом и корректной валидацией. Остальное докрутить можно. 

Будем пушить дальше.

**Reply by coder1coder** [reply]
`while (IsCodeReivewComplete == false)
Console.WriteLine("Please, can u review my code?");
`

**Reply by pingvin1308** [reply]
К субботе 27-го проверю все домашки, на прошедших выходных не успел xD

**Reply by pingvin1308** [reply]
Ты очень близок к тому как программы реализуются в продуктовой разработке. Весьма неплохо, есть ощущение, что ты уже пробовал веб разработку. Так как вижу репозитории и контроллеры)

**Reply by pingvin1308** [reply]
Лучше всегда указывать модификаторы доступа, вижу иногда опускаешь эти моменты

**Reply by pingvin1308** [reply]
Еще мелочь, лучше не использовать сокращения в названиях параметров.
Прикольно сделал менюшку)

**Reply by coder1coder** [reply]
> Ты очень близок к тому как программы реализуются в продуктовой разработке. Весьма неплохо, есть ощущение, что ты уже пробовал веб разработку. Так как вижу репозитории и контроллеры)

Есть грешок

> Лучше всегда указывать модификаторы доступа, вижу иногда опускаешь эти моменты
> Еще мелочь, лучше не использовать сокращения в названиях параметров.
> Прикольно сделал менюшку)

учту

**Comment by emedvedeva** [comment]
Вот моя домашка. Скидываю только для того, чтобы показать что делала я ее все таки. Не доделана. Поняла, что лучше использовать массивы и буду додумывать классы и интерфейсы. С ними что то не так)
https://github.com/emedvedeva/CsharpForBeginners/tree/main/Shop

**Reply by Z1pP** [reply]
Не вижу ваш sln файл для запуска проекта. Хочется потыкать и продебажить

**Reply by emedvedeva** [reply]
https://github.com/emedvedeva/CsharpForBeginners/tree/main/HomeTaskShop - вот сюда загрузила с солюшеном

**Reply by Z1pP** [reply]
Меня впечатлило использование обобщения. Я начал свой проект без них и теперь не знаю как их приспособить! 
Я бы попросил тебя отделить интерфейсы и классы, а то все так вместе. И очень прошу комментировать. Твоя работа выглядит грамотно. Удивлюсь, если никто не помогал за спиной)

**Reply by emedvedeva** [reply]
Меня муж уговаривал не пользоваться дженериками и интерфейсами. Говорит получилась "буханка-троллейбус"))) 
Приятно читать столь высокую оценку моего кода. 
По поводу комментариев - учту. Постараюсь комментировать.

**Reply by emedvedeva** [reply]
![4_1027326](https://user-images.githubusercontent.com/35913899/109385709-ed201700-7906-11eb-8dba-c5671b9555c1.jpg)

**Comment by alexrus79** [comment]
Здравствуйте.
Мою тоже посмотрите.
https://github.com/alexrus79/CsharpForBeginners/tree/main/MyFirstConsoleApp/HomeWork

**Reply by coder1coder** [reply]
[Здесь буду передавать свою точку зрения](https://github.com/alexrus79/CsharpForBeginners/issues)
ps: Ни одного интерфейса я не увидел

**Reply by alexrus79** [reply]
@coder1coder 
Спасибо. 
Я, если честно, не разобрался с интерфейсами. 
Роману об этом говорил. 
Он сказал скидывай что есть.

**Reply by pingvin1308** [reply]
Думаю  #region тут излишне, код выглядит немного спресованным)
Еще непонятно для чего ты реализовывал сеттеры у класса Product. По мне это сейчас выглядит как усложнение, лучше вместо этого сделать валидацию модели, но я не рассказывал как это делается. По этому лучше вообще убрать реализацию у сеттеров)

А еще класс Product выглядит очень божественным, тут слишком много информации

**Reply by pingvin1308** [reply]
Мне нравится твоя работа, она выглядит очень сложно. Нам будет где разгуляться на занятии в воскресенье. Я буду проводить так называемую работу над ошибками и показывать что и как можно было сделать

**Comment by blood-mif** [comment]
Добрый день! Догоняю вас как могу, пожалуйста, покритикуйте кто может
https://github.com/blood-mif/CsharpForBeginners/tree/main/Market

**Reply by pingvin1308** [reply]
Хорошо сделал, есть пару мелких замечаний

**Comment by Dmitriy-Nox** [comment]
Всем привет! Наконец нашел время для написания консольного магазином. Если будет время, посмотрите/прокомментируйте)))
https://github.com/Dmitriy-Nox/CcharpForBeginers/tree/main/TestGitHub/Shop

**Reply by pingvin1308** [reply]
Хорошо сделал, есть что дорабатывать. Жду обновления программы :)

**Reply by Dmitriy-Nox** [reply]
Залил новую версию. Исправил кое что по стилистике, постарался немного причесать код. Модель пока не трогал, до нее еще доберусь!

**Comment by eniluck** [comment]
Добрый день! Дописываю свою реализацию домашки. Осталось добавить товары и UI для их добавления в витрину.
https://github.com/eniluck/CsharpForBeginners/tree/master/Shop_homework15

**Reply by pingvin1308** [reply]
Молодец, все здорово сделал. Есть небольшие мелочи которые можно улучшить и поправить)

**Comment by 8bit-com** [comment]
https://github.com/8bit-com/Market.git Неделя у меня ушла на этот проект, есть ещё баги, но уже можно оценить

**Reply by eniluck** [reply]
Прикольно выглядит :) Только у меня редактирование витрины почему то не работает.

**Reply by 8bit-com** [reply]
У меня тоже)

**Reply by 8bit-com** [reply]
исправил, теперь работает

**Reply by pingvin1308** [reply]
Выглядит весьма недурно!

Из советов по доработке: 
Класс Shop выглядит немного перегруженным. Представь, что тебя попросили туда что нибудь добавить)
К примеру добавить возможность заказать товар с витрины через корзину. Насколько просто это будет делаться?
Попробуй разнести код класса Shop на структуры поменьше.

**Reply by Z1pP** [reply]
Согласен с Романом. Класс shop очень сильно завален кодом. Также я заметил "магические числа", которые встречаются с позиции курсора. А так, интересно, можно развивать )

**Reply by 8bit-com** [reply]
Это третья версия shop, раньше было хуже) Сейчас делаю сетевую, буду лучше оптимизировать

**Comment by FanTei** [comment]
https://github.com/FanTei/MyWork/tree/Main/Shop

**Reply by pingvin1308** [reply]
Названия коммитов следует делать более осознанными. Для этого нужно понимать какую работу ты сделал и что ты хочешь сейчас зафиксировать. На работе тебе не скажут спасибо за коммиты с описанием аа АА :)

**Comment by FanTei** [comment]
https://github.com/FanTei/MyWork/tree/Main/Shop
Переделал магазин

**Comment by IlyaEliseev** [comment]
https://github.com/IlyaEliseev/CsharpForBeginners/tree/main/Shop
Роман, привет. Посмотри как будет время.

**Reply by pingvin1308** [reply]
Посмотрел, в целом неплохое начало. Есть что дорабатывать, по этому предлагаю двигаться потихоньку. Как сделаешь правки по моим задачам в [Issues](https://github.com/IlyaEliseev/CsharpForBeginners/issues). Напиши комент сюда, а то вдруг нотификашка не придет :)

**Reply by IlyaEliseev** [reply]
Привет. Все поправил и оставил комментарии в Issues.

**Reply by IlyaEliseev** [reply]
Переделал проект

**Reply by pingvin1308** [reply]
Супер :)

**Reply by IlyaEliseev** [reply]
Добавил репозиторий.
